File: src/main/org/nlogo/agent/Topology3D.java
package org.nlogo.agent;

import org.nlogo.api.AgentException;

public interface Topology3D {
  double distanceWrap(double dx, double dy, double dz,
                      double x1, double y1, double z1,
                      double x2, double y2, double z2);

  double towardsPitchWrap(double dx, double dy, double dz);

  Patch getPatchAt(double xc, double yc, double zc) throws AgentException;

  AgentSet getNeighbors3d(Patch3D source);

  AgentSet getNeighbors6(Patch3D source);

  Patch getPNU(Patch3D source);

  Patch getPEU(Patch3D source);

  Patch getPSU(Patch3D source);

  Patch getPWU(Patch3D source);

  Patch getPNEU(Patch3D source);

  Patch getPSEU(Patch3D source);

  Patch getPSWU(Patch3D source);

  Patch getPNWU(Patch3D source);

  Patch getPND(Patch3D source);

  Patch getPED(Patch3D source);

  Patch getPSD(Patch3D source);

  Patch getPWD(Patch3D source);

  Patch getPNED(Patch3D source);

  Patch getPSED(Patch3D source);

  Patch getPSWD(Patch3D source);

  Patch getPNWD(Patch3D source);

  double observerZ();

  double wrapZ(double z);

  double shortestPathZ(double z1, double z2);
}

File: netlogo-gui/src/main/agent/Topology3D.scala
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.agent

import org.nlogo.api.AgentException

trait Topology3D {
  def distanceWrap(dx: Double, dy: Double, dz: Double,
                   x1: Double, y1: Double, z1: Double,
                   x2: Double, y2: Double, z2: Double): Double
  def towardsPitchWrap(dx: Double, dy: Double, dz: Double): Double
  @throws(classOf[AgentException])
  def getPatchAt(xc: Double, yc: Double, zc: Double): Patch
  def getNeighbors3d(source: Patch3D): IndexedAgentSet
  def getNeighbors6(source: Patch3D): IndexedAgentSet
  def getPNU(source: Patch3D): Patch
  def getPEU(source: Patch3D): Patch
  def getPSU(source: Patch3D): Patch
  def getPWU(source: Patch3D): Patch
  def getPNEU(source: Patch3D): Patch
  def getPSEU(source: Patch3D): Patch
  def getPSWU(source: Patch3D): Patch
  def getPNWU(source: Patch3D): Patch
  def getPND(source: Patch3D): Patch
  def getPED(source: Patch3D): Patch
  def getPSD(source: Patch3D): Patch
  def getPWD(source: Patch3D): Patch
  def getPNED(source: Patch3D): Patch
  def getPSED(source: Patch3D): Patch
  def getPSWD(source: Patch3D): Patch
  def getPNWD(source: Patch3D): Patch
  def observerZ: Double
  def wrapZ(z: Double): Double
  def shortestPathZ(z1: Double, z2: Double): Double
}
